Full version of EventSetup consumes system

Dr15Jones opened the pull request:
#### PR description:

The EventSetup consumes system now works with the ESGetToken to provide a more efficient data product lookup from a Record.

#### PR validation:

The code compiles and all framework related unit tests pass.

Dr15Jones committed:
Move templated functions to EventSetup

Moved the implementations of the templated functions from EventSetupImp to EventSetup. These functions are only needed by the users of EventSetup, not EventSetupImpl.
This sets up for further changes for EventSetup and records.

Dr15Jones committed:
The EventSetup system now enforces the transitionID

When using an ESGetToken, the EventSetup system makes sure that token was created for the given transition.
A transition can be e.g. BeginRun or Event in the ED system or is a given Record in the ES system.

Dr15Jones committed:
Fixed includes and forward declarations

Dr15Jones committed:
Detect if ESGetToken has been initialized

Dr15Jones committed:
Provide easy access to numbers of Transitions

Dr15Jones committed:
Use ESConsumesInfo

Dr15Jones committed:
Provide adapter for default_record to avoid need of ESHandle

Dr15Jones committed:
Record the EventSetup consumes info for ED modules

The information passed to the esConsumes call is now stored.

Dr15Jones committed:
Added ESRecordToProxyIndices class

The class is used to hold information obtained from all EventSetup Records which is needed for the upcoming change to ESGetToken.
Added methods to various EventSetup data structures to allow access to the needed information.

Dr15Jones committed:
Propagate ESRecordsToProxyIndices to all modules

ESRecordsToProxyIndices is used by the modules to translate the ESGetToken indices to information needed to do the actual ES product lookup.

Dr15Jones committed:
Fix message logger threshold in test

The test looks at the log to determine success. The log expects more than 100 messages to be given per category.
This problem was only found when compiling with '-O0 -g'. Apparently there is  bug in the message logger which causes the normal build to not properly enforce the limits.

Dr15Jones committed:
Fix sort order problem in ESRecordsToProxyIndices test

The test assumed that after a sort, the container would be in a particular entry order. This is not actually guaranteed since different compiler optimization levels can change the sort order.

Dr15Jones committed:
Added rvalue reference methods to DataKey

Dr15Jones committed:
Added explicit EventSetupsController::finishConfiguration

Instead of waiting for first synchronization, the configuration can be finalized explicitly.

Dr15Jones committed:
ESGetToken now uses indices

The ESGetToken now just contains indices, not the strings from an ESInputTag. These indices are used to lookup the index for DataProxies in the appropriate EventSetup Record. The proxy index is then used to directly call the Proxy.
The lookup tables from token index to proxy index are filled at begin job time.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26449/9254

- This PR adds an extra 296KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26449/9254/code-checks.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26449/9254/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

Dr15Jones committed:
Applied code-check change

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26449/9255

- This PR adds an extra 308KB to repository

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for master.

It involves the following packages:

FWCore/Framework
FWCore/Integration
FWCore/TestProcessor
FWCore/Utilities


@cmsbuild, @smuzaffar, @Dr15Jones can you please review it and eventually sign? Thanks.
@makortel, @felicepantaleo, @wddgit, @Martin-Grunewald this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented on FWCore/Utilities/interface/ESGetToken.h:
```diff
@@ -38,12 +40,29 @@ namespace edm {
+    constexpr ESGetToken<ESProduct,ESRecord>& operator=(ESGetToken<ESProduct,ESRecord> const&) noexcept = default;
+
+    constexpr unsigned int transitionID() const noexcept { return m_transitionID;}
+    constexpr bool isInitialized() const noexcept { return transitionID() != std::numeric_limits<unsigned int>::max();}
```
`EDGetToken` has `isUninitialized()`
https://github.com/cms-sw/cmssw/blob/7965eaafa4d3c9cbebf976d27f9ddad0bfbad27a/FWCore/Utilities/interface/EDGetToken.h#L47

makortel replied:
I actually think that `isInitialized()` would be more convenient to use, so maybe we should move `EDGetToken` to that instead?

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 59a73a301c4d17242fb191facdeca01976955ff3


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/de2816de91a7e625b5da6bd775cd1aebb6cbebe0
https://github.com/cms-sw/cmssw/commit/b544ff322951bd453ea5016d339c6f13429082a4
https://github.com/cms-sw/cmssw/commit/5b7b34f74d6b6ac8606f33d2428f9ac87e2203b7
https://github.com/cms-sw/cmssw/commit/f71ec216380755b2ca88f94d68722f12d2cc5488
https://github.com/cms-sw/cmssw/commit/87461208d41271d2a56c2042ad16e21a7c1d814b
https://github.com/cms-sw/cmssw/commit/e3f89fba2f34131c85795bfdc43b923cf98ec80e
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26449/34184/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26449/34184/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26449/34184/summary.html

I found follow errors while testing this PR

**Failed tests:  Build  HeaderConsistency  ClangBuild**

* **Build**:

I found compilation error when building: 

<pre>>> Compiling  /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_6_X_2019-04-11-2300/src/Mixing/Base/src/PileUp.cc 
>> Compiling  /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_6_X_2019-04-11-2300/src/Mixing/Base/src/BMixingModule.cc 
>> Compiling  /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_6_X_2019-04-11-2300/src/Mixing/Base/src/SecondaryEventProvider.cc 
In file included from /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_6_X_2019-04-11-2300/src/Mixing/Base/src/SecondaryEventProvider.cc:1:0:
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_6_X_2019-04-11-2300/src/Mixing/Base/src/SecondaryEventProvider.h: In member function 'void edm::SecondaryEventProvider::beginJob(const edm::ProductRegistry&)':
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_6_X_2019-04-11-2300/src/Mixing/Base/src/SecondaryEventProvider.h:28:87: error: no matching function for call to 'edm::WorkerManager::beginJob(const edm::ProductRegistry&)'
     void beginJob(ProductRegistry const& iRegistry) {workerManager_.beginJob(iRegistry);}
                                                                                       ^
In file included from /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_6_X_2019-04-11-2300/src/Mixing/Base/src/SecondaryEventProvider.h:4:0,
                 from /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_6_X_2019-04-11-2300/src/Mixing/Base/src/SecondaryEventProvider.cc:1:
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_6_X_2019-04-11-2300/src/FWCore/Framework/interface/WorkerManager.h:79:10: note: candidate: void edm::WorkerManager::beginJob(const edm::ProductRegistry&, const edm::eventsetup::ESRecordsToProxyIndices&)
</pre>

* **Clang**:

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/de2816de91a7e625b5da6bd775cd1aebb6cbebe0
https://github.com/cms-sw/cmssw/commit/b544ff322951bd453ea5016d339c6f13429082a4
https://github.com/cms-sw/cmssw/commit/5b7b34f74d6b6ac8606f33d2428f9ac87e2203b7
https://github.com/cms-sw/cmssw/commit/f71ec216380755b2ca88f94d68722f12d2cc5488
https://github.com/cms-sw/cmssw/commit/87461208d41271d2a56c2042ad16e21a7c1d814b
https://github.com/cms-sw/cmssw/commit/e3f89fba2f34131c85795bfdc43b923cf98ec80e
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26449/34184/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26449/34184/git-merge-result

cmsbuild commented:
Comparison not run due to Build errors (RelVals and Igprof tests were also skipped)

Dr15Jones committed:
Make PileUp compile with new EventSetup API

This code will compile and run using the non ESGetToken based
data product get calls. Additional work will be needed to support
consumes from the PileUp system.

cmsbuild commented:
The code-checks are being triggered in jenkins.

Dr15Jones commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26449/9262

- This PR adds an extra 472KB to repository

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #26449 was updated. @cmsbuild, @smuzaffar, @civanch, @Dr15Jones, @mdhildreth can you please check and sign again.

cmsbuild commented:
-1

Tested at: 5d18d70666b41142be8b933133a4a280d770e927

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26449/34188/summary.html

I found follow errors while testing this PR

**Failed tests:  Build  HeaderConsistency  ClangBuild**

* **Build**:

I found compilation warning when building:  See details on the summary page.

* **Clang**:

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

cmsbuild commented:
Comparison not run due to Build errors (RelVals and Igprof tests were also skipped)

Dr15Jones committed:
Removed unnecessary braces in intialization

Fixes a clang warning.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26449/9263

- This PR adds an extra 472KB to repository

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #26449 was updated. @cmsbuild, @smuzaffar, @civanch, @Dr15Jones, @mdhildreth can you please check and sign again.

cmsbuild commented:
+1
Tested at: b8f4a54b3749991b03f3131979acf999754eb459
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26449/34189/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26449/34189/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3142783
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3142585
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

Dr15Jones commented:
+1

civanch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
@Dr15Jones as I would like to converge on pre4 by tomorrow, is it perhaps better to integrate this large development after that?

Dr15Jones commented:
Delaying this integration  is fine.

Dr15Jones committed:
Improved exception message when using ESGetToken

Distinguish between an invalid token an a token to a missing Proxy.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26449/9346

- This PR adds an extra 288KB to repository

cmsbuild commented:
Pull request #26449 was updated. @cmsbuild, @smuzaffar, @civanch, @Dr15Jones, @mdhildreth can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 466ed9057b319dd03ddcbe716335979672104f8a
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26449/34265/summary.html

cmsbuild commented:
Comparison job queued.

Dr15Jones commented:
+1

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26449/34265/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3142783
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3142585
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

civanch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

wddgit commented on FWCore/Framework/interface/ESProducer.h:
```diff
@@ -152,19 +157,23 @@ namespace edm {
                          TReturn (T ::* iMethod)(const TRecord&),
                          const TArg& iDec,
                          const es::Label& iLabel = {}) {
+      const auto id = static_cast<unsigned int>(numberOfFactories());
```
Possibly I am just not understanding something here. What does the numberOfFactories have to do with the transitionID getting passed into the Callback?

Dr15Jones replied:
Each factory generates a proxy for one and only one transition (i.e. EventSetup record). So the number of factories is just a way to identify which proxy is being called.

wddgit replied:
What is worrying me is id is being passed into the constructor for the Callback immediately below this point (you have to scroll far to the right to see it in my browser). That is supposed to be a transition ID in the Callback constructor, not a Proxy identifier. Where id is used below that is probably good, but where it is passed into the Callback just seems incorrect. What if there are 10 DataProxy's? Then the value would be outside the range of the enum for transition IDs.

wddgit replied:
One alternative would be to propagate the transition ID as an argument through all the various get* function calls in the stack that leads all the way back to the EventSetup get function call.

Dr15Jones replied:
@wddgit You are thinking that 'transitionID' here is of type 'edm::TransitionID'. It is not. For EDProducers the numeric value I use to keep track of 'transitionID' is the one from 'edm::TransitionID' since EDProducers have 'callbacks' (i.e. virtual functions) for only those transitions. For ESProducers, a 'transitionID' is an identifier for which 'setWhatProduced' is associated with which Callback. I.e. the value of 'transitionID' here is just local to that particular ESProducer. It is say which 'ESConsumesCollector' gathered the information for that given Callback.

wddgit replied:
Now I understand. Thanks.

fabiocos commented:
+1
